src/plugins/lua/contextal.lua

```diff
@@ -0,0 +1,319 @@
+  end
+end
+
+local function submit_cb(task)
+  if cache_context then
+    redis_cache.cache_get(task,
+        task:get_digest(),
+        cache_context,
+        settings.cache_timeout,
+        submit,
+        cache_hit
+    )
+  else
+    submit(task)
+  end
+end
+
+local function action_cb(task)
+  local work_id = task:get_mempool():get_variable('contextal_work_id', 'string')
+  if not work_id then
+    rspamd_logger.err(task, 'no work id found in mempool')
+    return
+  end
+
+  local function http_callback(err, code, body, hdrs)
+    if err then
+      rspamd_logger.err(task, 'http error: %s', err)
+      maybe_defer(task)
+      return
+    end
+    if code ~= 200 then
+      rspamd_logger.err(task, 'bad http code: %s', code)
+      maybe_defer(task)
+      return
+    end
+    local parser = ucl.parser()
+    local _, parse_err = parser:parse_string(body)
+    if parse_err then
+      rspamd_logger.err(task, 'cannot parse JSON: %s', err)
+      maybe_defer(task)
+      return
+    end
+    local obj = parser:get_object()
+    return process_actions(task, obj, false)
+  end
+
+  rspamd_http.request({
+      task = task,
+      url = settings.actions_url .. work_id,
+      callback = http_callback,
+      timeout = settings.http_timeout,
+      gzip = settings.gzip,
+      keepalive = settings.keepalive,
+      no_ssl_verify = settings.no_ssl_verify,
+  })
+end
+
+local function set_url_path(base, path)
+  local slash = base:sub(#base) == '/' and '' or '/'
+  return base .. slash .. path
+end
+
+settings = lua_util.override_defaults(settings, opts)
+
+local res, err = config_schema:transform(settings)
+if not res then
+  rspamd_logger.warnx(rspamd_config, 'plugin %s is misconfigured: %s', N, err)
+  local err_msg = string.format("schema error: %s", res)
+  lua_util.config_utils.push_config_error(N, err_msg)
+  lua_util.disable_module(N, "failed", err_msg)
+  return
+end
+
+contextal_actions = lua_util.list_to_hash(contextal_actions)
+for _, k in ipairs(settings.custom_actions) do
+  contextal_actions[k] = true
+end
+
+if not settings.base_url then
+  if not (settings.submit_url and settings.actions_url) then
+    rspamd_logger.err(rspamd_config, 'no URL configured for contextal')
+    lua_util.disable_module(N, 'config')
+    return
+  end
+else
+  if not settings.submit_url then
+    settings.submit_url = set_url_path(settings.base_url, 'api/v1/submit')
+  end
+  if not settings.actions_url then
+    settings.actions_url = set_url_path(settings.base_url, 'api/v1/actions/')
+  end
+end
+
+redis_params = lua_redis.parse_redis_server(N)
+if redis_params then
+  cache_context = redis_cache.create_cache_context(redis_params, {
+      cache_prefix = settings.cache_prefix,
+      cache_ttl = settings.cache_ttl,
+      cache_format = 'json',
+      cache_use_hashing = false
+  })
+end
+
+rspamd_config:register_symbol({
+  name = settings.submission_symbol,
+  priority = lua_util.symbols_priorities.top,
+  type = 'prefilter',
```

Comment:
Interesting, why do we need that on prefilter stage? Maybe we should do other filtering first to allow, for example, whitelisting or early greylist/ratelimit to avoid expensive checks?

Reply:
The idea is merely to try execute it early which is a likely misuse.

If `request_ttl` is set (it is by default so far), and that is less than `task_timeout`, then probably we should use that to set a timer instead of using a postfilter.